src/qwircl/core.cljs

```diff
@@ -33,34 +51,108 @@
   ; nothing yet
   state)
 
-(defn to-position [event]
-  ; x is easy, but for y we need to take the header
-  ; into account
-  (let [x (int (/ (:x event) size))
-        r (:y event)]
-    (if (<= r header)
-      {:in :hand :coordinates [x]}
-      {:in :grid :coordinates [x (int (/ (- r header) size))]}))) 
+(defn inside-dimensions? [xp yp {:keys [x y w h]}]
+  (and 
+   (<= x xp (+ x w))
+   (<= y yp (+ y h))))
+
+(defn get-clicked [x y]
+  (first (filter #(inside-dimensions? x y (% dimensions)) [:hand :grid :submit :undo])))
+
+(defn translate-hand [xp hand]
+  (let [x (int (/ (- xp (first header-translation)) size))]
+    (when (and (<= 0 x) (< x (count hand)))
+      [x])))
+
+(defn translate-grid [x y]
+  [(int (/ x size)) (int (/ (- y header) size))])
+
+(defn translate-event [state {:keys [x y]}]
+  (condp = (get-clicked x y)
+    :hand (when-let [h (translate-hand x (get-in state [:my :hand]))] 
+            {:action :hand-clicked :clicked h})
+    :grid {:action :grid-clicked :clicked (translate-grid x y)}
+    :submit {:action :submit}
+    :undo {:action :undo}
+    {:x x :y y}))
+
+(defn empty-location? [[x y] grid]
+  (let [location (get-in grid [x y])]
+    (and 
+     (nil? (:color location))
+     (nil? (:shape location)))))
+
+(defn valid-play? [state clicked]
+  (and 
+   (empty-location? clicked (:grid state))
+   true))
+
+;; push-down automaton for managing state for
+;; picking and playing tiles
+(defn run-pda 
+  [{{:keys [s hand positions] :as pda} :pda :as state} {:keys [action clicked]}]
+  (condp = s
+    :initial 
+    (cond
+      (= action :hand-clicked) {:s :picking :hand [clicked] :positions []}
+      :else {:s :initial})
+    :picking 
+    (cond
+      (and 
+       (= action :hand-clicked)
```

Comment:
I think this method seems like a multimethod dispatching on `[state action]` where the default implementation is to return `state` unmodified.

Reply:
Got it.